Detectors/CTP/workflow/src/RawToDigitConverterSpec.cxx

```diff
@@ -0,0 +1,157 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+#include <string>
+#include "FairLogger.h"
+#include "CommonDataFormat/InteractionRecord.h"
+//#include "Framework/ConfigParamRegistry.h"
+//#include "Framework/ControlService.h"
+#include "Framework/WorkflowSpec.h"
+#include "DetectorsRaw/RDHUtils.h"
+#include "DPLUtils/DPLRawParser.h"
+#include "CTPWorkflow/RawToDigitConverterSpec.h"
+
+using namespace o2::ctp::reco_workflow;
+
+void RawToDigitConverterSpec::init(framework::InitContext& ctx)
+{
+}
+
+void RawToDigitConverterSpec::run(framework::ProcessingContext& ctx)
+{
+  mOutputDigits.clear();
+  std::map<o2::InteractionRecord, CTPDigit> digits;
+  const gbtword80_t bcidmask = 0xfff;
+  gbtword80_t pldmask;
+  using InputSpec = o2::framework::InputSpec;
+  using ConcreteDataTypeMatcher = o2::framework::ConcreteDataTypeMatcher;
+  using Lifetime = o2::framework::Lifetime;
+  //mOutputHWErrors.clear();
+  std::vector<InputSpec> filter{InputSpec{"filter", ConcreteDataTypeMatcher{"CTP", "RAWDATA"}, Lifetime::Timeframe}};
+  o2::framework::DPLRawParser parser(ctx.inputs(), filter);
+  uint32_t payloadCTP;
+  for (auto it = parser.begin(); it != parser.end(); ++it) {
+    auto rdh = it.get_if<o2::header::RAWDataHeader>();
+    auto triggerOrbit = o2::raw::RDHUtils::getTriggerOrbit(rdh);
+    auto linkCRU = o2::raw::RDHUtils::getLinkID(rdh); // 0 = IR, 1 = TCR
+    if (linkCRU == o2::ctp::CRULinkIDIntRec) {
+      payloadCTP = o2::ctp::NIntRecPayload;
+    } else if (linkCRU == o2::ctp::CRULinkIDClassRec) {
+      payloadCTP = o2::ctp::NClassPayload;
+    } else {
+      LOG(ERROR) << "Unxpected  CTP CRU link:" << linkCRU;
+    }
+    pldmask = 0;
+    for (uint32_t i = 0; i < payloadCTP; i++) {
+      pldmask[12 + i] = 1;
+    }
+    // TF in 128 bits words
+    gsl::span<const uint8_t> payload(it.data(), it.size());
+    gbtword80_t gbtWord = 0;
+    gbtword80_t remnant = 0;
+    uint32_t size_gbt = 0;
+    int wordCount = 0;
+    for (auto payloadWord : payload) {
+      if (wordCount == 15) {
+        wordCount = 0;
+      } else if (wordCount > 9) {
+        wordCount++;
+      } else if (wordCount == 9) {
+        wordCount++;
+        std::vector<gbtword80_t> diglets;
```

Comment:
Instead of having dynamic object creation in a tight loop it would be better to define it outside of all loops and call `clear` here.